extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -105,8 +108,20 @@ public QueryDefinition makeQueryDefinition(
 
     final int firstStageNumber = Math.max(minStageNumber, queryDefBuilder.getNextStageNumber());
     final WindowOperatorQuery queryToRun = (WindowOperatorQuery) originalQuery.withDataSource(dataSourcePlan.getNewDataSource());
-    final int maxRowsMaterialized;
 
+    // Get segment granularity from query context, and create ShuffleSpec and RowSignature to be used for the final window stage.
+    final Granularity segmentGranularity = QueryKitUtils.getSegmentGranularityFromContext(jsonMapper, queryToRun.getContext());
+    final ClusterBy finalWindowClusterBy = QueryKitUtils.clusterByWithSegmentGranularity(ClusterBy.none(), segmentGranularity);
+    ShuffleSpec finalWindowStageShuffleSpec = resultShuffleSpecFactory.build(finalWindowClusterBy, false);
+    if (Objects.equals(segmentGranularity, Granularities.ALL)) {
+      finalWindowStageShuffleSpec = MixShuffleSpec.instance();
+    }
```

Comment:
This seems wrong. We should still be using the `resultShuffleSpecFactory` to create the final shuffle spec. If the windowing stage is an input to the segment generator, then the window processor would create one large partition, while we expect the segments to be sized correctly. Am I missing the reason for this additional check?

Reply:
```java
    if (Objects.equals(segmentGranularity, Granularities.ALL)) {
      finalWindowStageShuffleSpec = MixShuffleSpec.instance();
    }
```

Removing the above lines causes failures when we have multiple workers (`maxNumTasks` > 2).

Queries fail with:
```
Caused by: org.apache.druid.java.util.common.IAE: Cannot shuffle with spec [MaxCountShuffleSpec{clusterBy=ClusterBy{columns=[], bucketByCount=0}, partitions=4, aggregate=false}] and nil clusterBy
```

Thoughts?

Reply:
I've pushed some logic changes as per offline discussion. Please have a look, will close this comment as it's outdated now. Thanks!